### PR TITLE
feat: support drop_last=False during validation

### LINKED ISSUE #605: Add support for drop_last=False for validation dataloader
**Is your feature request related to a problem? Please describe.**

Currently, the last validation batch is dropped
https://github.com/NVIDIA-NeMo/RL/blob/3f6d52fc884d6ffc1ab881fd1cb1853cd6ef9eff/nemo_rl/algorithms/sft.py#L165 which affects the accuracy of the results, especially for small datasets.

**Describe the solution you'd like**

From a user perspective, add a new config parameter `policy.drop_last_validation` that can be set to either True or False.

**Additional context**

This will require changes to data sharding, for example here https://github.com/NVIDIA-NeMo/RL/blob/f4150bf76681a4aac70fb83529505ddae5bcfb8c/nemo_rl/distributed/batched_data_dict.py#L289

### LINKED ISSUE #958: Why drop_last in sft validation dataloader?
In the [SFT dataloader](https://github.com/NVIDIA-NeMo/RL/blob/38e9ef10be7424a03af669a89c001e8f25d3740c/nemo_rl/algorithms/sft.py#L148), the validation data loader has `drop_last=True`, is that a design choice or a bug? Since we want the validation loss be representative of all validation samples, would it make sense to set it to False, or the code logic would break?

### PR DESCRIPTION
**Pull request by ashors1**
# What does this PR do ?
Adds a new key `val_drop_last` to sft, dpo and rm to allow users to run validation on the full eval set. This is enabled by padding the final validation batch to a multiple of `mbs * dp_size`. 

# Issues
closes #605
closes #958 

# Usage
* **You can potentially add a usage example below**

```python
# Add a code snippet demonstrating how to use this
```

# Before your PR is "Ready for review"
**Pre checks**:
- [ ] Make sure you read and followed [Contributor guidelines](/NVIDIA-NeMo/RL/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you run the unit tests and functional tests locally? Visit our [Testing Guide](/NVIDIA-NeMo/RL/blob/main/docs/testing.md) for how to run tests
- [ ] Did you add or update any necessary documentation? Visit our [Document Development Guide](/NVIDIA-NeMo/RL/blob/main/docs/documentation.md) for how to write, build and test the docs.

# Additional Information
* ...

### CODE REVIEW

File: nemo_rl/algorithms/utils.py [hunk 1]
```diff
@@ -265,3 +267,51 @@ def get_tokenizer(
...
+        )
+        # Pad input_lengths
+        batch["input_lengths"] = torch.cat(
+            [
+                batch["input_lengths"],
+                batch["input_lengths"][-1].unsqueeze(0).repeat(min_padding),
+            ]
+        )
+        # Pad token_mask
+        batch["token_mask"] = torch.cat(
```
Comment (jveronvialard) [inline nemo_rl/algorithms/utils.py]: token_mask is not always present (e.g. for RM), see https://github.com/NVIDIA-NeMo/RL/blob/de918124243289c7da20042b40c346ee8301eefc/nemo_rl/data/datasets.py#L299

File: nemo_rl/algorithms/dpo.py [hunk 2]
```diff
@@ -343,16 +347,24 @@ def validate_one_dataset(
         print(f"▶ Starting validation at step {step} for `{dataset_name}` set..")
 
         val_metrics = defaultdict(lambda: 0.0)
-        num_valid_batches = 0
+        sum_num_valid_samples = 0
         for batch_idx, val_batch in enumerate(
             add_ref_logprobs_to_data(val_dataloader, policy, master_config, is_val=True)
         ):
+            # Check if we need to pad the final batch to make it divisible by micro_batch_size * dp_size
```
Comment (jveronvialard) [inline nemo_rl/algorithms/dpo.py]: Why not do it in `AllTaskProcessedDataset` instead? To avoid the multiple calls to `maybe_pad_last_batch()` across different algorithm files

File: nemo_rl/algorithms/sft.py [hunk 3]
```diff
@@ -285,14 +288,17 @@ def validate(
                     " This is likely because there were no valid samples."
                 )
             else:
-                val_metrics["val_loss"] += float(val_results["loss"])
-                num_valid_batches += 1
+                num_tokens = (
```
Comment (jveronvialard) [inline nemo_rl/algorithms/sft.py]: nit: rename `num_tokens` to `num_valid_tokens` and `num_valid_tokens` to `sum_num_valid_tokens`

File: nemo_rl/algorithms/dpo.py [hunk 4]
```diff
@@ -364,19 +376,24 @@ def validate_one_dataset(
 
             else:
                 for k, v in val_results["all_mb_metrics"].items():
-                    if k in {"lr", "wd", "global_valid_seqs", "global_valid_toks"}:
+                    samples_in_this_batch = val_batch["sample_mask"].sum()
```
Comment (jveronvialard) [inline nemo_rl/algorithms/dpo.py]: How about consolidating this implementation with that of rm.py?

File: nemo_rl/algorithms/dpo.py (hunk 2 above)
Comment (peri044) [reply nemo_rl/algorithms/dpo.py]: @ashors1 I have same question. I was wondering if you encountered any drawbacks making this change in collate functions of these the `AllTaskProcessedDataset` objects based on our discussion

File: nemo_rl/algorithms/dpo.py [hunk 5]
```diff
@@ -343,16 +347,24 @@ def validate_one_dataset(
         print(f"▶ Starting validation at step {step} for `{dataset_name}` set..")
 
         val_metrics = defaultdict(lambda: 0.0)
-        num_valid_batches = 0
+        sum_num_valid_samples = 0
         for batch_idx, val_batch in enumerate(
             add_ref_logprobs_to_data(val_dataloader, policy, master_config, is_val=True)
         ):
+            # Check if we need to pad the final batch to make it divisible by micro_batch_size * dp_size
+            if val_batch.size < val_batch_size * 2:
```
Comment (peri044) [inline nemo_rl/algorithms/dpo.py]: `val_batch_size * 2` can be subtle - consider using `mbs_last_val = 2` which generalizes across rm.py and dpo.py

File: nemo_rl/algorithms/utils.py (hunk 1 above)
Comment (ashors1) [reply nemo_rl/algorithms/utils.py]: Fixed, thanks!

File: nemo_rl/algorithms/dpo.py (hunk 2 above)
Comment (ashors1) [reply nemo_rl/algorithms/dpo.py]: Good point, this slipped my mind. I'll try to move this into a more central location

File: nemo_rl/algorithms/dpo.py (hunk 5 above)
Comment (ashors1) [reply nemo_rl/algorithms/dpo.py]: Sorry, I don't think I follow. Could you elaborate on this suggestion?

File: nemo_rl/algorithms/dpo.py (hunk 2 above)
Comment (ashors1) [reply nemo_rl/algorithms/dpo.py]: I looked into this briefly but after a bit more consideration I think it makes sense to keep the implementation where it is. We need to know the data parallel size to determine the number of padding samples, but the collate function has no understanding of DP size. I also think that by adding `maybe_pad_last_batch` to the shared `utils.py` file, we avoid much of the redundant code that might arise from having to add this padding logic to each algorithm individually. Please let me know what you think of this!

Comment (jveronvialard) [reply nemo_rl/algorithms/dpo.py]: When calling setup() for a given algorithm, why not initialize the cluster and policy first, then the train and validation dataloaders? This way, you'll have access to the DP size through the policy (as you currently do) and could add a new parameter to the collate_fn e.g. `pad_to_multiple_of`. If you prefer your current implementation, that's fine with me too, I agree that the changes are not too involved. Maybe also add a comment in `add_ref_logprobs_to_data`?

File: examples/configs/sft.yaml [hunk 6]
```diff
@@ -8,8 +8,9 @@ sft:
   val_period: 10
   val_batches: 8
   val_global_batch_size: 32
-  val_micro_batch_size: 1
+  val_micro_batch_size: 4
```
Comment (jveronvialard) [inline examples/configs/sft.yaml]: Why this change?

File: nemo_rl/algorithms/dpo.py (hunk 5 above)
Comment (peri044) [reply nemo_rl/algorithms/dpo.py]: Consider storing this as an argument `mbs_last_val=2` and using `val_batch.size < val_batch_size * mbs_last_val` for easier error debugging. This `mbs_last_val` would be 1 in sft.py and 2 in dpo/rm.py.
Also, I saw in discussions and in this PR https://github.com/NVIDIA-NeMo/RL/pull/977, that dataset is being refactored. I see `PreferenceDataset` and `ResponseDataset`. So maybe `mbs_last_val` can be an attribute of these classes for a cleaner implementation - I don't have complete understanding of this PR but just a thought.

File: nemo_rl/algorithms/dpo.py (hunk 2 above)
Comment (ashors1) [reply nemo_rl/algorithms/dpo.py]: If you're okay with it, I think I'd prefer to keep this logic in the algorithms themselves. Changing the size of the batch within the collate function feels slightly unusual to me, and since the logic depends on attributes within the policy, I feel like it's more intuitive to handle the logic in a place that has access to both the batch and the policy. 

>Maybe also add a comment in add_ref_logprobs_to_data?

Makes sense, will do

File: nemo_rl/algorithms/dpo.py (hunk 5 above)
Comment (ashors1) [reply nemo_rl/algorithms/dpo.py]: I like the idea of this approach but I'm not sure how useful this will be with the current state of the code, since the logic for when to pad a batch isn't shared across algorithms. Also, I think the name `mbs_last_val` might be slightly misleading since this is actually the factor we multiply the batch sizes by to get the true batch size used in the forward pass. 

What do you think about revisiting this idea in a follow-up PR?

File: examples/configs/sft.yaml (hunk 6 above)
Comment (ashors1) [reply examples/configs/sft.yaml]: This isn't related to this PR, so I'll remove it. It just makes validation slightly more performant

File: nemo_rl/algorithms/dpo.py (hunk 2 above)
Comment (jveronvialard) [reply nemo_rl/algorithms/dpo.py]: Ok, sounds good to me too

File: nemo_rl/algorithms/dpo.py [hunk 7]
```diff
@@ -165,7 +168,7 @@ def setup(
             ],
             add_loss_mask=True,
         ),
-        drop_last=True,
+        drop_last=dpo_config["val_drop_last"],
```
Comment (jveronvialard) [inline nemo_rl/algorithms/dpo.py]: Should this be reverted?

File: nemo_rl/algorithms/dpo.py [hunk 8]
```diff
@@ -255,6 +258,12 @@ def add_ref_logprobs_to_data(dataloader, policy, master_config, is_val=False):
                 else master_config["policy"]["train_micro_batch_size"] * 2
             )
 
+            # when running validation with drop_last=False, we might end up with a partial batch.
+            # In this case, we pad the batch to the next multiple of micro_batch_size * dp_size.
+            dp_size = policy.sharding_annotations.get_axis_size("data_parallel")
+            if batch.size % (dp_size * micro_batch_size) != 0:
+                batch = maybe_pad_last_batch(batch, dp_size, micro_batch_size)
```
Comment (jveronvialard) [inline nemo_rl/algorithms/dpo.py:272]: Should you also assert `is_val=True`?

File: nemo_rl/algorithms/rm.py [hunk 9]
```diff
@@ -307,14 +307,19 @@ def validate_one_dataset(
         dict_val_metrics = defaultdict(list)
         num_valid_batches = 0
         for batch_idx, val_batch in enumerate(val_dataloader):
+            # Check if we need to pad the final batch to make it divisible by micro_batch_size * dp_size
```
Comment (jveronvialard) [inline nemo_rl/algorithms/rm.py]: nit
```suggestion
            # When running validation with drop_last=False, we might end up with a partial batch.
            # In this case, we pad the batch to the next multiple of micro_batch_size * dp_size.
```

File: nemo_rl/algorithms/sft.py [hunk 10]
```diff
@@ -269,13 +269,16 @@ def validate(
 
             # update multimodal data
             val_data.update(cat_and_padded.get_multimodal_dict(as_tensors=False))
+            if val_data.size < val_batch_size:
```
Comment (jveronvialard) [inline nemo_rl/algorithms/sft.py:274]: nit
```suggestion
            # When running validation with drop_last=False, we might end up with a partial batch.
            # Check if we need to pad the final batch to make it divisible by micro_batch_size * dp_size.
            if val_data.size < val_batch_size:
```

File: nemo_rl/algorithms/dpo.py (hunk 8 above)
Comment (ashors1) [reply nemo_rl/algorithms/dpo.py:272]: Good point, done

File: nemo_rl/algorithms/dpo.py (hunk 7 above)
Comment (ashors1) [reply nemo_rl/algorithms/dpo.py]: Oh yes good catch, fixed

File: nemo_rl/algorithms/utils.py [hunk 11]
```diff
@@ -265,3 +267,52 @@ def get_tokenizer(
         processor.name_or_path = tokenizer.name_or_path
 
     return tokenizer if processor is None else processor
+
+
+def maybe_pad_last_batch(batch: dict, dp_size: int, mbs: int) -> dict:
+    min_padding = (math.ceil(batch.size / (mbs * dp_size)) * mbs * dp_size) - batch.size
```
Comment (peri044) [inline nemo_rl/algorithms/utils.py:283]: Consider adding a docstring to this function

File: nemo_rl/algorithms/dpo.py (hunk 5 above)
Comment (peri044) [reply nemo_rl/algorithms/dpo.py]: This sounds good

**Review by jveronvialard — DISMISSED** [review]

File: nemo_rl/algorithms/utils.py (hunk 11 above)
Comment (ashors1) [reply nemo_rl/algorithms/utils.py:283]: Good point, done!

File: examples/configs/dpo.yaml [hunk 12]
```diff
@@ -7,6 +7,7 @@ dpo:
   val_global_batch_size: 8
   val_micro_batch_size: 1
   val_at_start: true
+  val_drop_last: false
```
Comment (terrykong) [inline examples/configs/dpo.yaml]: discussed offline, prob don't need another toggle and can just re-use `val_batches` since these overlap

File: nemo_rl/algorithms/dpo.py [hunk 13]
```diff
@@ -361,22 +373,44 @@ def validate_one_dataset(
                     "No validation metrics were collected for this batch."
                     " This is likely because there were no valid samples."
                 )
-
             else:
-                for k, v in val_results["all_mb_metrics"].items():
-                    if k in {"lr", "wd", "global_valid_seqs", "global_valid_toks"}:
```
Comment (terrykong) [inline nemo_rl/algorithms/dpo.py:367]: does this change drop these keys from our metrics?
```
"lr", "wd", "global_valid_seqs", "global_valid_toks"
```

File: nemo_rl/algorithms/rm.py [hunk 14]
```diff
@@ -307,14 +307,20 @@ def validate_one_dataset(
...
+            if val_batch.size < val_batch_size * 2:
+                dp_size = policy.sharding_annotations.get_axis_size("data_parallel")
+                val_batch = maybe_pad_last_batch(val_batch, dp_size, val_mbs * 2)
+
             ## just run model fwd
             val_results = policy.train(
                 val_batch,
                 loss_fn,
                 eval_mode=True,
                 ## NOTE: we double the batch size here because each preference example corresponds to a pair of
```
Comment (terrykong) [inline nemo_rl/algorithms/rm.py]: should this comment be updated? maybe just move above this line and do an assert that
```
assert val_batch.sze == val_batch_size * 2, f"...."
```

File: nemo_rl/algorithms/utils.py [hunk 15]
```diff
@@ -265,3 +267,62 @@ def get_tokenizer(
         processor.name_or_path = tokenizer.name_or_path
 
     return tokenizer if processor is None else processor
+
+
+def maybe_pad_last_batch(batch: dict, dp_size: int, mbs: int) -> dict:
```
Comment (terrykong) [inline nemo_rl/algorithms/utils.py:272]: can you add a unit test for this?

File: nemo_rl/algorithms/dpo.py (hunk 13 above)
Comment (ashors1) [reply nemo_rl/algorithms/dpo.py:367]: yes, but I don't think "lr" and "wd" need to be logged for validation. I'll make sure we still log "global_valid_seqs" and "global_valid_toks"

File: nemo_rl/algorithms/rm.py (hunk 14 above)
Comment (ashors1) [reply nemo_rl/algorithms/rm.py]: It won't always be true that `assert val_batch.size == val_batch_size * 2`, right? For the last validation batch, `val_batch.size` should just be equal to a multiple of `mbs * 2 * dp_size`, but that should be true by construction

File: nemo_rl/algorithms/utils.py (hunk 15 above)
Comment (ashors1) [reply nemo_rl/algorithms/utils.py:272]: Yep, done

**Comment by ashors1** [comment]
> thanks @ashors1, left some comments. Does grpo.py need this as well?

Nope, since GRPO's validation loop is slightly different from the other algorithms, it already has `drop_last=False`

File: nemo_rl/algorithms/dpo.py (hunk 13 above)
Comment (terrykong) [reply nemo_rl/algorithms/dpo.py:367]: Makes sense. I didn't see `global_valid_seqs` anywhere. Did it get dropped?

Also, it went from mean -> sum. Was that intentional?

File: nemo_rl/algorithms/rm.py (hunk 14 above)
Comment (terrykong) [reply nemo_rl/algorithms/rm.py]: I just meant asserting after padding, which should be the case right? But on second thought, I think it's fine to not assert. Unit tests cover the "correctness" here.

File: nemo_rl/algorithms/dpo.py (hunk 13 above)
Comment (ashors1) [reply nemo_rl/algorithms/dpo.py:367]: `global_valid_seqs` and `num_valid_samples` are basically the same thing, but I guess there is value in both because `global_valid_seqs` tells us how many sequences were passed to `policy.train` and `num_valid_samples` tells us how many train examples were run (i.e. in the case of DPO, `global_valid_seqs = 2 * num_valid_samples`). I'll add it back.

>Also, it went from mean -> sum. Was that intentional?

Yes, averaging across the batches was a bug. We want to sum the number of samples across the batches rather than averaging

Comment (ashors1) [reply nemo_rl/algorithms/dpo.py:367]: > Also, it went from mean -> sum. Was that intentional?

Correction: we want to sum across batches but average across microbatches within a batch. Looking into this now

Comment (jveronvialard) [reply nemo_rl/algorithms/dpo.py:367]: Sorry, why do we want to average global_valid_seqs and global_valid_toks across microbatches?

Comment (ashors1) [reply nemo_rl/algorithms/dpo.py:367]: The `global_*` variables describe the number of samples/tokens in the entire batch. Those values are used internally to ensure the loss computation is independent of the microbatch size for a fixed batch.

Comment (jveronvialard) [reply nemo_rl/algorithms/dpo.py:367]: Got it, thanks

**Review by jveronvialard — APPROVED** [review]

**Review by terrykong — DISMISSED** [review]
